### PR TITLE
core: add support for setting CPUAffinity= to special "numa" value

### LINKED ISSUE #13248: RFE: CPUAffinity=NUMA to simplify unification of CPU and Memory resource contraints
From #3097: thanks for adding [NUMAPolicy](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#NUMAPolicy=) and [NUMAMask](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#NUMAMask=).

A `CPUAffinity=NUMA` would be a useful extension to align `CPUAffinity` to the same node as memory.

**Is your feature request related to a problem? Please describe.**

The ability to set [NUMAMask](#12764) to constrain (per NUMAPolicy) memory allocated to a node usually has a concurrent thought to constrain CPUAffinity to the same NUMAMask nodes as the memory.

The determination of which CPUs are associated with the same memory nodes used in NUMAMask is left as a fragile exercise to the sysadmin leading to faults especially when one settings is changed without another.

**Describe the solution you'd like**

A setting of `CPUAffinity=NUMA` could concisely equate the memory and CPU constraints without the possibility of error.

**Describe alternatives you've considered**

Manually entering `CPUAffinity=` explicit list of CPUs. - Fragile.

`CPUAffinity=NUMA:{numanode1}, NUMA:{numanode1}`, less fragile, however duplication of the list of nodes in `NUMAMask` and not sure what practical use it is to constrain a service to number of CPUs on one node, but memory from two nodes. There might be a cause - if there is this could still be implemented later.

As a libnuma function the start of this is [numa_node_to_cpus](https://linux.die.net/man/3/numa_node_to_cpus), returning a set of cpus.

### PR DESCRIPTION
**Pull request by msekletar**
systemd will automatically derive CPU affinity mask from NUMA node
mask.

Fixes #13248

### CODE REVIEW

**Comment by mrc0mmand** [comment]
Don't mind the `TEST-46-HOMED` fail, filed as #14755.

File: src/core/dbus-execute.c [hunk 1]
```diff
@@ -220,7 +222,19 @@ static int property_get_cpu_affinity(
         assert(reply);
         assert(c);
 
-        (void) cpu_set_to_dbus(&c->cpu_set, &array, &allocated);
+        if (!c->cpu_affinity_from_numa)
```
Comment (keszybz) [inline src/core/dbus-execute.c]: I'd prefer to remove the negation and switch the if and else branches...

File: src/core/dbus-execute.c [hunk 2]
```diff
@@ -220,7 +222,19 @@ static int property_get_cpu_affinity(
...
+                (void) cpu_set_to_dbus(&c->cpu_set, &array, &allocated);
+        else {
+                int r;
+                _cleanup_(cpu_set_reset) CPUSet s = {};
+
+                r = numa_to_cpu_set(&c->numa_policy, &s);
+                if (r < 0)
+                        return r;
+
+                (void) cpu_set_to_dbus(&s, &array, &allocated);
```
Comment (keszybz) [inline src/core/dbus-execute.c]: cpu_set_to_dbus could be pulled out from the two branches.

File: src/core/dbus-execute.c [hunk 3]
```diff
@@ -1767,6 +1782,22 @@ int bus_exec_context_set_transient_property(
 
                 return 1;
 
+        } else if (streq(name, "CPUAffinityFromNUMA")) {
+                bool q;
+
+                r = sd_bus_message_read_basic(message, 'b', &q);
```
Comment (keszybz) [inline src/core/dbus-execute.c:1789]: `sd_bus_message_read_basic` takes `int*` for `b`.

File: src/core/dbus-execute.c [hunk 4]
```diff
@@ -1767,6 +1782,22 @@ int bus_exec_context_set_transient_property(
...
+                if (r < 0)
+                        return r;
+
+                if (!UNIT_WRITE_FLAGS_NOOP(flags)) {
+                        if (q) {
+                                c->cpu_affinity_from_numa = q;
+                                unit_write_settingf(u, flags, name, "%s=%s", "CPUAffinity", "numa");
+                        }
+                }
```
Comment (keszybz) [inline src/core/dbus-execute.c:1796]: Maybe drop those outer {} ?

File: src/core/execute.c [hunk 5]
```diff
@@ -3012,6 +3012,33 @@ static int exec_parameters_get_cgroup_path(const ExecParameters *params, char **
...
+        int r;
+        _cleanup_(cpu_set_reset) CPUSet s = {};
+
+        assert(c);
+        assert(ret);
+
+        if (!c->numa_policy.nodes.set) {
+                log_warning("Can't derive CPU affinity mask from NUMA mask because NUMA mask is not set, ignoring");
+                return 0;
+        }
```
Comment (keszybz) [inline src/core/execute.c:3034]: Is this warning actually needed? If the numa mask is not set, the process gets all cpus. I think this should also cause the cpuaffinity mask to be silently not set (with just a log_debug message).

File: src/core/execute.c [hunk 6]
```diff
@@ -3309,11 +3336,27 @@ static int exec_child(
...
         }
 
-        if (context->cpu_set.set)
-                if (sched_setaffinity(0, context->cpu_set.allocated, context->cpu_set.set) < 0) {
-                        *exit_status = EXIT_CPUAFFINITY;
-                        return log_unit_error_errno(unit, errno, "Failed to set up CPU affinity: %m");
-                }
+        if (context->cpu_set.set || context->cpu_affinity_from_numa) {
+                _cleanup_(cpu_set_reset) CPUSet converted_cpu_set = {};
+                const CPUSet *cpu_set = NULL;
```
Comment (keszybz) [inline src/core/execute.c]: unnecessary initialization

File: src/core/execute.c [hunk 7]
```diff
@@ -3309,11 +3336,27 @@ static int exec_child(
...
+                        if (r < 0) {
+                                *exit_status = EXIT_CPUAFFINITY;
+                                return log_unit_error_errno(unit, r, "Failed to derive CPU affinity mask from NUMA mask: %m");
+                        }
+
+                        cpu_set = &converted_cpu_set;
+                } else if (context->cpu_set.set)
+                        cpu_set = &context->cpu_set;
+
+                if (cpu_set)
```
Comment (keszybz) [inline src/core/execute.c]: This is always true?

File: src/core/dbus-execute.c [hunk 8]
```diff
@@ -220,7 +222,19 @@ static int property_get_cpu_affinity(
         assert(reply);
         assert(c);
 
-        (void) cpu_set_to_dbus(&c->cpu_set, &array, &allocated);
+        if (!c->cpu_affinity_from_numa)
+                (void) cpu_set_to_dbus(&c->cpu_set, &array, &allocated);
+        else {
+                int r;
+                _cleanup_(cpu_set_reset) CPUSet s = {};
```
Comment (keszybz) [inline src/core/dbus-execute.c]: Lennart like `int r` to be last...

File: src/core/load-fragment.c [hunk 9]
```diff
@@ -1329,14 +1329,26 @@ int config_parse_exec_cpu_affinity(const char *unit,
                                    void *data,
                                    void *userdata) {
 
+        int r;
         ExecContext *c = data;
```
Comment (keszybz) [inline src/core/load-fragment.c:1332]: Here too, please put the definition of `r` last.

File: src/core/execute.c [hunk 10]
```diff
@@ -3012,6 +3012,33 @@ static int exec_parameters_get_cgroup_path(const ExecParameters *params, char **
...
+        r = numa_to_cpu_set(&c->numa_policy, &s);
+        if (r < 0)
+                return r;
+
+        cpu_set_reset(ret);
+
+        return cpu_set_add_all(ret, &s);
+}
+
+bool exec_context_get_cpu_affinity_from_numa(const ExecContext *c) {
```
Comment (kkuehlz) [inline src/core/execute.c:3045]: What is the point of putting this in a function? The only place it’s called already asserts the context

Comment (msekletar) [reply src/core/execute.c:3045]: Function is used in D-Bus property declaration,

```
static BUS_DEFINE_PROPERTY_GET(property_get_cpu_affinity_from_numa, "b", ExecContext, exec_context_get_cpu_affinity_from_numa);
```

**Comment by msekletar** [comment]
I've pushed the new version. @keszybz PTAL

File: src/core/dbus-execute.c [hunk 11]
```diff
@@ -1770,6 +1783,21 @@ int bus_exec_context_set_transient_property(
...
 
+        } else if (streq(name, "CPUAffinityFromNUMA")) {
+                int q;
+
+                r = sd_bus_message_read_basic(message, 'b', &q);
+                if (r < 0)
+                        return r;
+
+                if (!UNIT_WRITE_FLAGS_NOOP(flags))
+                        if (q) {
```
Comment (keszybz) [inline src/core/dbus-execute.c]: Hmm, I think this needs to be written unconditionally. It is possible that the setting is set to true somewhere in existing configuration, and then setting it to false needs to be written to override that.

File: src/core/execute.c [hunk 12]
```diff
@@ -3309,11 +3336,26 @@ static int exec_child(
                 }
         }
 
-        if (context->cpu_set.set)
-                if (sched_setaffinity(0, context->cpu_set.allocated, context->cpu_set.set) < 0) {
+        if (context->cpu_set.set || context->cpu_affinity_from_numa) {
+                _cleanup_(cpu_set_reset) CPUSet converted_cpu_set = {};
+                const CPUSet *cpu_set = NULL;
```
Comment (keszybz) [inline src/core/execute.c]: This doesn't need to be initalized.

File: src/core/execute.c [hunk 13]
```diff
@@ -3309,11 +3336,26 @@ static int exec_child(
...
+
+                if (context->cpu_affinity_from_numa) {
+                        r = exec_context_cpu_affinity_from_numa(context, &converted_cpu_set);
+                        if (r < 0) {
+                                *exit_status = EXIT_CPUAFFINITY;
+                                return log_unit_error_errno(unit, r, "Failed to derive CPU affinity mask from NUMA mask: %m");
+                        }
+
+                        cpu_set = &converted_cpu_set;
+                } else if (context->cpu_set.set)
```
Comment (keszybz) [inline src/core/execute.c]: Not that it matters, but this `if (...)` could be dropped.

File: src/core/execute.c [hunk 14]
```diff
@@ -3012,6 +3012,33 @@ static int exec_parameters_get_cgroup_path(const ExecParameters *params, char **
...
+        if (!c->numa_policy.nodes.set) {
+                log_debug("Can't derive CPU affinity mask from NUMA mask because NUMA mask is not set, ignoring");
+                return 0;
+        }
+
+        r = numa_to_cpu_set(&c->numa_policy, &s);
+        if (r < 0)
+                return r;
+
+        cpu_set_reset(ret);
```
Comment (keszybz) [inline src/core/execute.c:3040]: This is not necessary, right?

Comment (msekletar) [reply src/core/execute.c:3040]: I wanted to make sure that returned CPUSet contains only CPUs that correspond to NUMA mask hence I am resetting it before I add there list of CPUs derived from NUMA mask.

**Comment by msekletar** [comment]
@keszybz I've force pushed the branch. PTAL

File: src/core/execute.c (hunk 14 above)
Comment (keszybz) [reply src/core/execute.c:3040]: Oh right, this is not about `s` but about `ret`. Never mind.

**Comment by msekletar** [comment]
CentOS CI failed because of the test timeout in TEST-33-CLEAN-UNIT. According to @mrc0mmand the test failure is mostly likely flake caused by nested KVM. 

Ubuntu CI failed (AFAICT) due to following timeout in 551/777 fuzz-dns-packet:oss-fuzz-13422:address  TIMEOUT 30.17 s 

Both failures seem unrelated so I think it should be OK to merge this. @keszybz PTAL.

**Comment by mrc0mmand** [comment]
The intermediate results of CentOS CI show that all integration tests passed (now it's running the last testsuite for systemd-networkd), so it was indeed a flake.

**Comment by grooverdan** [comment]
Thank you @msekletar for writing this feature. Thanks @keszybz for reviewing it.

**Comment by shibumi** [comment]
@msekletar  do you have an example configuration for this? I would like to create tmpfs mounts based on the number of available NUMA nodes. 

for example, if I have 8 NUMA nodes, I would like to automatically set up tmpfs mounts with these options here:

```
rw,relatime,size=400G,mpol=prefer:<NUMA NODE>" "tmpfs<NODE ID>" "/FOO/tmpfs<NODE ID>/FOOBAR"
```

How would I do this with your change?

**Pull request merged**